async parser handler support

jerch opened the pull request:
Not being able to use async functions in parser handlers is an unfortunate limitation of the current parser implementation. This PR tries to come up with a solution to enable async support while keeping the parser performant.

TODO:
- [x] async parser support for:
  - [x] CSI handler
    - [x] basic impl
    - [x] test cases
  - [x] ESC handler
    - [x] basic impl
    - [x] test cases
  - [x] OSC handler
    - [x] basic impl
    - [x] test cases
  - [x] DCS handler
    - [x] basic impl
    - [x] test cases
- [x] async/sync mixed test cases on inputhandler
- [x] high level API tests with async handlers
- [x] deprecate/remove `writeSync` (refactor core tests) - needs note in release doc
- [x] proper `reset` handling on parser
- [x] simplify/cleanup handler interface mess of the parser (see #3223)
- [x] cleanup async codebase
- [x] update parser hooks docs --> created https://github.com/xtermjs/xtermjs.org/issues/136

Shall fix #3218.

jerch commented:
The early commit can be tested in the browser by marking SGR as async - simply replace:
https://github.com/xtermjs/xterm.js/blob/74ea558bfc8ed2900fd90166803fc323ca254cfe/src/common/InputHandler.ts#L2301

with
```typescript
  public async charAttributes(params: IParams): Promise<void> {
```
(Note that our test will not run anymore with that, as they heavily rely on `writeSync`.)

jerch commented:
The last commit enables to test async handlers in our inputhandler transparently, which helps to spot regressions while messing around with random async handlers and testing the perf behavior.

The next thingy to reshape it is `writeSync`, after that the core tests should also work with async handlers.

jerch commented:
@Tyriar The last commit deprecates `writeSync` as I it wont work with async handlers anymore.

**Reminder**: this should be noted later on in release notes in case someone relies on it.

jerch commented:
Done with the basic parser preparations. After some stack handling cleanup the async test case (`ls -lR /usr/lib` with `SGR` marked as async) takes only slightly longer (~1.4 times or 29 vs. 21 MB/s), guess this wont get any better.

Next steps:
- Implement parser tests for CSI and ESC mixing all sort of sync and async handlers.
- Carry on to OSC and DCS async support. As written earlier, this is much more involved, as these handlers are fed by 3 methods from the parser via separate subparsers (`start` | `put` | `end`). Imho `start` can always be treated as sync, much in the sense of a constructor. `put` is more tricky - from my initial perspective to get proper flow control from async handlers in general, async support in `put` would be crucial as here huge data amounts may arrive. Not sure yet to go that path, a simpler impl would just treat the whole sequence as being blocked on the `finish` call (which again needs a fixed mem limit during `put` to avoid OOM and some bailout logic for async workers).

jerch commented:
@Tyriar @mofux 
The PR is ready for a first review.

Tyriar reviewed (changes requested):
Mostly looks good, no concerns about stability/correctness considering the amount of testing coverage/changes.

jerch commented on src/common/CoreTerminal.ts:
```diff
@@ -125,7 +125,17 @@ export abstract class CoreTerminal extends Disposable implements ICoreTerminal {
+   * be used anymore. If you need blocking semantics on data input consider
+   * `write` with a callback instead.
+   *
+   * @deprecated Unreliable, will be removed soon.
```
Yeah, we also had many tests with `writeSync`, but those were easy replaced by a `writeP` implementation: https://github.com/xtermjs/xterm.js/blob/26bf1ab840ba9f0ffcc3a2196d4053552b880b55/src/browser/TestUtils.test.ts#L24-L26

I used the same `methodP` idea for testing the methods along the async callstack: https://github.com/xtermjs/xterm.js/blob/26bf1ab840ba9f0ffcc3a2196d4053552b880b55/src/common/InputHandler.test.ts#L50-L56

This is a bit more cumbersome to get done right, thus my warnings all over the place not to call those methods directly anymore. Note that it always was dangerous to directly inject chunks at those lower levels while `Terminal.write` still holds data (will mess up parser states even in sync mode), but now with async it is even more dangerous as it might break the proper continuation. Lets hope that ppl dont dismantle the callstack and call the lower methods directly. Also I dont know of any TS way to prevent that, is there an access pattern like `friend` in C++?

Final note on `writeSync` - I did not remove it yet, because it will keep working as before as long as there are no async handlers hooked into the parser. It even keeps working with async handlers if pending data does not trigger any async action (thats what I meant with promise islands - you can still sail in sync water normally at high speed, but now there are promise shallow regions where the boat can crash if not used with `writeP`). Initially I thought about marking async handlers statically, so the parser would have a "map" of promise regions, but that turned out as not being helpful, as it cannot be decided prehand without actually parsing the data. Thus I went with the simpler handling of "whenever a promise is returned, it is an async handler" (effectively sailing blindfolded).

jerch commented on src/common/InputHandler.ts:
```diff
@@ -460,10 +460,57 @@ export class InputHandler extends Disposable implements IInputHandler {
+   * Note: Never call this directly for a running terminal instance in production.
+   * Always use `Terminal.write`, which provides in-band blocking and correct exection order.
+   */
+  public parse(data: string | Uint8Array, promiseResult?: boolean): void | Promise<boolean> {
```
Well I dont see much gain from variant 1. and 2. beside easier dealing with a terminal breakdown during tests. Main reason - the proper continuation is placed at `WriteBuffer` level in `then`, which `InputHandler.parse` is not aware of, thus the right execution order cannot be continued from `InputHandler.parse` directly anymore. I'd favour a more restrictive access pattern here (like `friend` in C++), but dont know how to do something like that in TS.

In theory during a normal terminal run with valid data in the write buffer the callstack of `Terminal.write --> InputHandler.parse --> Parser.parse --> (Osc|DcsParser.end) --> async handler` must not be "sidecalled", whenever an async handler got paused, as it has to be resumed from the very top.

Tyriar replied:
Maybe I didn't explain enough. I think the above would give a compile error if `parse(data, false)` was called, and it would return `Promise<boolean>` when `parse(data, true)` is called (instead of `void | Promise<boolean>`). Just a little thing that could potentially catch bugs by indicating what is valid.

jerch replied:
Not sure if I fully understand your idea - do you want to conclude from provided arguments the return type, like narrowing it down? That is not possible, as the fact, whether a promise gets returned, depends on these two rules:
- an async handler is registered for a sequence XY (and will not be skipped by an earlier handler)
- AND `data` contains sequence XY

It never depends on the arguments of `parse`, all these variants can return `void | Promise<boolean>`:
```typescript
parse(data: string | Uint8Array): void | Promise<boolean>;
parse(data: string | Uint8Array, promiseResult: true): Promise<boolean>;
parse(data: string | Uint8Array, promiseResult: false): Promise<boolean>;

// which reduces to
parse(data: string | Uint8Array, promiseResult?: boolean): void | Promise<boolean>;
```

In fact the promise return value and the result argument are responsible for the two-sided handling of the stalled callstack as pausing/resuming:
- promise return value
  - ascending path of stalled callstack (on pause)
  - stack progress is saved at each call level individually
  - fast unwinding by direct return jumps
  - promise bubbles up to top level to enable micro- or macrotask exection
-  micro- or macrotask exection, eventually our promise gets fullfilled and we grab its return value
- promiseReturn: 
  - descending path of stalled callstack (on resume)
  - rewind stack progress from stack saves on each level
  - bubbles down to occurrence of promise creation, continues innermost resume with return value (indicating whether next handler should be probed as well)
  - back to sync processing

The only assumption that could be made is for two consecutive `parse` calls:
- call returns a promise - then next call has to be called with the awaited result of that promise in `promiseResult` for proper continuation and execution order (we are between pause/resume)
- `parse` calls finishs without returning a promise - we are in sync mode, nothing to wait for before we can process the next chunk

Technically the whole handling is a quite explicit way of coroutines as one could write them in C (where it can be expressed even more elegant with static vars :see_no_evil:). The different call requirements for sync vs. async is an unfortunate side effect of this explicit coroutine implementation. But as noted earlier, it was the fastest I was able to find (under the assumption of only few to none async handlers ever being registered). True that the full promisified call chain has much simpler call signatures, but it also runs 4 times slower.

> ... Just a little thing that could potentially catch bugs by indicating what is valid.

Since we can only derive the proper call signature at runtime by knowing the result of the last call, imho TS's static type system cannot help here much.

Sorry for the lengthy explanation, there is still a high chance I misses your idea here.

jerch commented on src/common/InputHandler.ts:
```diff
@@ -460,10 +460,57 @@ export class InputHandler extends Disposable implements IInputHandler {
+   *   execution stopped at async handler, stack saved, continue with
+   *   same chunk and the promise resolve value as `promiseResult` until the method returns `undefined`
+   *
+   * Note: Never call this directly for a running terminal instance in production.
+   * Always use `Terminal.write`, which provides in-band blocking and correct exection order.
```
Yeah lets deescalate wording here a bit. I placed those freaking out warnings in the beginning all over the place, because sidecalling those methods will create havoc to execution order and terminal state. Youre right, it is very unlikely ppl will call into those functions directly.

jerch commented on src/common/input/WriteBuffer.ts:
```diff
@@ -37,8 +37,9 @@ export class WriteBuffer {
-  constructor(private _action: (data: string | Uint8Array) => void) { }
+  constructor(private _action: (data: string | Uint8Array, promiseResult?: boolean) => void | Promise<boolean>) { }
 
+  // FIXME: does not work that way anymore with async handlers!!!
```
To be removed.

jerch commented on src/common/input/WriteBuffer.ts:
```diff
@@ -76,16 +77,64 @@ export class WriteBuffer {
 
-  protected _innerWrite(): void {
-    const startTime = Date.now();
+  protected _innerWrite(d: number = 0, promiseResult: boolean = true): void {
```
Gonna use a selfspeaking name, will also place some explanation in the doc strings. (It is an optimization to allow multiple async microtasks to run within one bigger `setTimeout` frame).

jerch commented on src/common/input/WriteBuffer.ts:
```diff
@@ -76,16 +77,64 @@ export class WriteBuffer {
+        //   ? r => setTimeout(() => this._innerWrite(0, r))
+        //   : r => this._innerWrite(startTime, r);
+
+        result.then(continuation, err => { setTimeout(() => { throw err; }); continuation(true); });
```
Hmm, pulling `continuation` like that in front will still throw the exception way too late, as it would continue before showing the error. Furthermore errors might stack up, if  `continuation` produces more errors. I think the only way to keep this halfway in sync with the current "band-progress" would be to catch error first before scheduling any continuation. Gonna change that.

jerch commented on src/common/parser/DcsParser.ts:
```diff
@@ -79,20 +83,46 @@ export class DcsParser implements IDcsParser {
     if (!this._active.length) {
       this._handlerFb(this._ident, 'UNHOOK', success);
     } else {
+      let handlerResult: any = false;
```
Yeah, seems thats a restriction of `instanceof` in TS? I was not able to narrow it down prehand. Any idea?

jerch commented on src/common/parser/EscapeSequenceParser.ts:
```diff
@@ -435,9 +444,37 @@ export class EscapeSequenceParser extends Disposable implements IEscapeSequenceP
+    handlers: ResumableHandlersType,
+    handlerPos: number,
+    transition: number,
+    chunkPos: number): void {
```
Yeah, thats how I normally structure things myself. But the "Format Document" has a different idea about it :smile_cat:, which I sometimes use if I changed several indentation blocks.

jerch commented on typings/xterm.d.ts:
```diff
@@ -1501,6 +1501,23 @@ declare module 'xterm' {
+   * In general custom handlers should also operate in sync mode wherever
+   * possible to keep the parser fast.
+   * Still the exposed interfaces allow to register async handlers by returning
+   * a `Promise<boolean>`. Here the parser will pause input processing until
+   * the promise got resolved or rejected (in-band blocking). This "full stop"
```
Ah thats a very good idea - well we cannot savely resume by a fixed time constraint, as the paused handler still might returned after 5mins and does its nasty alterations. But at least a console warning would be helpful for devs to see why things are stuck. Gonna add that.

jerch commented on src/common/input/WriteBuffer.ts:
```diff
@@ -76,16 +77,64 @@ export class WriteBuffer {
     while (this._writeBuffer.length > this._bufferOffset) {
       const data = this._writeBuffer[this._bufferOffset];
+
+      if (result = this._action(data, promiseResult)) {
```
Changed to const value assigment before the condition check.

jerch commented on src/common/parser/DcsParser.ts:
```diff
@@ -79,20 +83,46 @@ export class DcsParser implements IDcsParser {
+      }
+      if (!fallThrough && handlerResult === false) {
+        for (; j >= 0; j--) {
+          if ((handlerResult = this._active[j].unhook(success)) !== false) {
```
Simplified all conditional asigments.

jerch commented:
@Tyriar Thx for the review! guess I addressed most points. Also if you have an idea how to apply a more restrictive access pattern like `friend` in C++ I am all ears.

jerch commented:
@Tyriar I changed the continuation of faulty handlers to `false` to give the default handler a chance to still get called afterwards. I think thats the better default behavior, as most ppl will hook on already handled sequences, thus the sequence should keep bubbling to our default handler, even if the custom async handler error'ed out.

Some open questions:

_Do we need a second timeout with a hard rejection?_
Currently we have a timeout warning after 5s, that fires only once. Imho this is enough for a careful dev to spot limitations of his handler implementation and to ensure himself, that the promise **always** gets resolved/rejected in a **timely fashion**. Not so for a sloppy dev - the code might have dangling promise endpoints due to improper branch testing and such. Here the terminal would block forever, if such an endpoint got triggered.
To avoid that hard blocking scenario in production environments later on we could place a second timeout on a pending promise and let it fail hard/reject after that second timeout. This would keep the terminal interactive/responsive, although the terminal state might be messed up, if the custom handler already altered something prior promise creation (or in the executor itself). But it kinda raises "realtime" questions about guaranteed execution times.


_Better async stack introspection needed?_
It is somewhat tricky to get better introspection information on the blocking handler, mainly because of the scattered stack saves and the way we register custom handlers through anonymous in-between functions. The scattered stack is not a big deal, it could be collected by a sync descent into the callstack, still the information is somewhat useless without proper marking of the handler function itself. If we want better introspection here, we would have to store more information along the registering path.

I dont see both aspects as showstoppers of the actual PR and suggest to implement them in later PRs if we encounter issues around blocking async handlers and need more failstate handling to overcome them.

Tyriar commented:
> Also if you have an idea how to apply a more restrictive access pattern like friend in C++ I am all ears.

I normally stick "friend classes" in the same file and don't export them. Closest I've come up with for TS.

Tyriar commented:
> Do we need a second timeout with a hard rejection?

The warning is good, recovering is probably a good idea too so the terminal doesn't just break because of some addon.

> Better async stack introspection needed?

👍 to defer, not sure we need this but I haven't tried to implement an async handler.

Tyriar commented on src/common/InputHandler.ts:
```diff
@@ -460,10 +464,70 @@ export class InputHandler extends Disposable implements IInputHandler {
   }
 
-  public parse(data: string | Uint8Array): void {
+  /**
+   * Async parse support.
+   */
+  private _parseStack = {
+    paused: false,
+    cursorStartX: 0,
+    cursorStartY: 0,
+    decodedLength: 0,
+    position: 0
+  };
```
Pull IParseStack into an interface and move var declaration to top?

Tyriar commented on src/common/InputHandler.ts:
```diff
@@ -460,10 +464,70 @@ export class InputHandler extends Disposable implements IInputHandler {
+  }
+  private _logSlowResolvingAsync(p: Promise<boolean>): void {
+    // log a limited warning about an async taking too long
+    if ((this._logService as any)._logLevel <= LogLevel.WARN) {
```
We can expose a `_logService.logLevel` since we need it here

jerch replied:
Yepp good idea.

Tyriar commented on src/common/parser/DcsParser.ts:
```diff
@@ -79,20 +83,47 @@ export class DcsParser implements IDcsParser {
   }
 
-  public unhook(success: boolean): void {
+  private _stack = {
```
Interface and prop up top?

Tyriar commented on src/common/parser/OscParser.ts:
```diff
@@ -137,12 +119,18 @@ export class OscParser implements IOscParser {
     }
   }
 
+  private _stack = {
```
Interface and prop up top? Can we share this interface with the other parsers?

jerch replied:
:+1: Yupp, OSC and DCS parser already share a common type base.

Tyriar reviewed (approved):
Lgtm 🚀

jerch commented:
@Tyriar I pressed the merge button :sweat_smile:. Since you mentioned above the explicit usage of `writeSync` in vscode - plz drop me a note if you encounter issues with a promise based replacement there.
